From bc670280b2683763c15e5e8500d5af987a47fa9a Mon Sep 17 00:00:00 2001
From: aegnux-contributor <contributor@example.invalid>
Date: Sun, 18 Oct 2026 05:04:55 +0300
Subject: [PATCH 12/18] dzn: enable the Vulkan memory model

Accept OpMemoryModel Vulkan (VulkanMemoryModel capability) in the SPIR-V to NIR
options, and report the KHR_vulkan_memory_model feature bits.
---
 src/microsoft/spirv_to_dxil/dxil_spirv_nir.c | 2 ++
 src/microsoft/vulkan/dzn_device.c            | 7 ++++---
 2 files changed, 6 insertions(+), 3 deletions(-)

diff --git a/src/microsoft/spirv_to_dxil/dxil_spirv_nir.c b/src/microsoft/spirv_to_dxil/dxil_spirv_nir.c
index 73cfbb9..a30040f 100644
--- a/src/microsoft/spirv_to_dxil/dxil_spirv_nir.c
+++ b/src/microsoft/spirv_to_dxil/dxil_spirv_nir.c
@@ -79,6 +79,8 @@ spirv_caps = {
    .Float64 = true,
    .Tessellation = true,
    .PhysicalStorageBufferAddresses = true,
+   .VulkanMemoryModel = true,
+   .VulkanMemoryModelDeviceScope = true,
 };
 
 static const struct spirv_to_nir_options
diff --git a/src/microsoft/vulkan/dzn_device.c b/src/microsoft/vulkan/dzn_device.c
index c3e2d7f..61f51ae 100644
--- a/src/microsoft/vulkan/dzn_device.c
+++ b/src/microsoft/vulkan/dzn_device.c
@@ -114,6 +114,7 @@ dzn_physical_device_get_extensions(struct dzn_physical_device *pdev)
       .KHR_driver_properties                 = true,
       .KHR_dynamic_rendering                 = true,
       .KHR_external_memory                   = true,
+      .KHR_vulkan_memory_model               = true,
       .EXT_depth_clip_enable                 = true,
       .KHR_external_semaphore                = true,
 #ifdef _WIN32
@@ -779,9 +780,9 @@ dzn_physical_device_get_features(const struct dzn_physical_device *pdev,
       .bufferDeviceAddress                = pdev->shader_model >= D3D_SHADER_MODEL_6_6,
       .bufferDeviceAddressCaptureReplay   = false,
       .bufferDeviceAddressMultiDevice     = false,
-      .vulkanMemoryModel                  = false,
-      .vulkanMemoryModelDeviceScope       = false,
-      .vulkanMemoryModelAvailabilityVisibilityChains = false,
+      .vulkanMemoryModel                  = true,
+      .vulkanMemoryModelDeviceScope       = true,
+      .vulkanMemoryModelAvailabilityVisibilityChains = true,
       .shaderOutputViewportIndex          = false,
       .shaderOutputLayer                  = false,
       .subgroupBroadcastDynamicId         = true,
-- 
2.43.0

